lib/spack/spack/util/compression.py

```diff
@@ -282,38 +307,108 @@ def decompressor_for(path, extension=None):
     # compression. If no lzma support, use tools available on systems
-    # 7zip on Windows and the xz tool on Unix systems.
     if re.match(r"xz", extension):
         return _lzma_decomp
 
-    # Catch tar.xz/tar.Z files here for Windows
-    # as the tar utility on Windows cannot handle such
-    # compression types directly
-    if ("xz" in extension or "Z" in extension) and sys.platform == "win32":
-        return _win_compressed_tarball_handler
-
     return _untar
 
 
+def _determine_py_decomp_archive_strategy(extension):
+    """Returns appropriate python based decompression strategy
+    based on extension type"""
+    # Only rely on Python decompression support for gz
+    if re.match(r"gz", extension):
+        return _py_gunzip
+
+    # Only rely on Python decompression support for bzip2
+    if re.match(r"bz", extension):
+        return _py_bunzip
+
+    # Python does not have native support
+    # of any kind for .Z files. In these cases,
+    # we rely on 7zip, which must be installed outside
+    # of spack and added to the PATH or externally detected
+    if re.match(r"Z", extension):
+        return _unZ
+
+    # Only rely on Python decompression support for xz
+    if re.match(r"xz", extension):
+        return _py_lzma
+
+    return None
+
+
+def decompressor_for_win(extension):
+    """Returns a function pointer to appropriate decompression
+    algorithm based on extension type and Windows specific considerations
+
+    Windows natively vendors *only* tar, no other archive/compression utilities
+    So we must rely exclusively on Python module support for all compression
+    operations, tar for tarballs and zip files, and 7zip for Z compressed archives
+    and files as Python does not provide support for the UNIX compress algorithm
+
+    Args:
+        path (str): path of the archive file requiring decompression
+        extension (str): extension
+    """
+
+    # Windows native tar can handle .zip extensions, use standard
+    # unzip method
+    if re.match(r"zip", extension):
+        return _unzip
+
+    # Windows vendors no native decompression tools, attempt to derive
+    # python based decompression strategy, if this is None, type is
+    # some sort of archive, compressed or other
+    decompressor = _determine_py_decomp_archive_strategy(extension)
+    if decompressor:
+        return decompressor
+    # reaching here means `path` is either a compressed archive or a tarball
+    # Windows does not natively provide *any* decompression utilities
+    # and only provides tar for archive extraction
+    # so we must decompress the archive with python modules and then hand things
+    # off to Windows native tar.exe
+
+    # if extension is standard tarball, invoke Windows native tar
+    if re.match(r"tar$", extension):
+        return _untar
+
+    extension = compression_ext_from_compressed_archive(extension)
+    decompressor = _determine_py_decomp_archive_strategy(extension)
```

Comment:
I disagree with some of this. If we perform this check a single time, we're either going to have to add anther check to determine if the extension we're looking at is a compressed tarball to decide whether to hand things off to `_win_compressed_tarball_handler` or a vanilla decompression method or we're just going to always blindly hand things off to that method, which is conceptually incorrect and might result in confusing stack traces/is inconsistent with the rest of the changes suggested which are primarily along conceptual modeling lines. Either way we're making an "extra call" or modeling incorrectly.
The `tar` change is fine.

Reply:
I'm unclear where the following would go wrong:

```
if re.match(r"zip", extension):
    return _unzip

extension = compression_ext_from_compressed_archive(extension)
decompressor = _determine_py_decomp_archive_strategy(extension)
if decompressor:
    return decompressor

if re.match(r"tar$", extension):
    return _untar

raise Exception(...)
```

In other words, if `re.match(r"tar$", extension)` finds a match, I don't see how `_determine_py_decomp_archive_strategy(extension)` would return something (other than `None`).

If that still seems wrong, could you explain the expected results for `foo.tar`, `foo.tar.gz`, `foo.gzip`, and `foo.tbz`? Ultimately, regardless of how (or whether this changes) it might be useful to have that in the docstring of this function.

Reply:
Sorry, I was not clear. The `tar` check change is one of the parts I do agree with. It absolutely makes the most sense to check for and return that prior to checking for any type of decompression.
I also agree with the error reporting change.
My issue is with the 
> * Think this should only call _determine_py_decomp_archive_strategy once (just start out by expanding the extension if it's possibly a contraction like .tgz)
suggestion.
We need to call this twice. Once on the original extension, and once on the expanded extension if the original doesn't return anything useful. We _could_ only call this once, but in that case, we'd either need to perform a check to see if the extension is a compressed tarball or not, or to allow the `_win_compressed_tarball_handler` handle both normal vanilla decompression _and_ compressed tarballs, which, while the method is set up to do so for robustness reasons, is not really conceptually accurate, which seems like it would be a problem considering that many of the other suggestions here are along primarily conceptual modeling lines.

Reply:
> We need to call this twice. Once on the original extension, and once on the expanded extension if the original doesn't return anything useful. 

I was thinking that if you start by calling `compression_ext_from_compressed_archive(extension)` to "normalize" it, then that would avoid having to call twice, but perhaps I misunderstood how that would work; I should say in that case that having such a normalizing function would be useful (which I assume would be simple - if that sounds like more than a few lines, I'm probably not understanding some underlying complexity).

Reply:
Right, so at that point in the code we know it's not a vanilla tarball, a zip file, or a vanilla .Z. We then start by "normalizing" the extension, so something like `tgz` becomes `tar.gz`, and things like `tar.gz` or `gz` stay the same. We then have a "normalized" ext where we don't have to think about the contraction extensions. We can then run the `_determine_py_decomp_archive_strategy(extension)` method on that extension, which gives us the compression type, `.gz` for all cases above. At this point, and this is where my issue arises, we would just be passing that derived decompressor to `_win_compressed_tarball_handler`, however the extension could be `gz` not `tar.gz` so passing it to the tarball handler as opposed to `_py_gunzip` doesn't really make a lot of conceptual sense, and in fact in addition to not making sense conceptually, without the conceptually redundant check for a tar extension being present I added to that handler, this approach would cause that to fail when tar tries to extract a file that is not a tarball. 
It's not really a normalization issue.